### PR TITLE
docs(Integrations): P2P-Persistence integration

### PR DESCRIPTION
**Pull request by deblasis**
Short description about P2P - Persistence integration and also scaffolding for similar documentation.

Origin: https://github.com/pokt-network/pocket/issues/271

CC: @andrewnguyen22

### CODE REVIEW

File: integrations/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+Without going too much into technical details, its main features for the purposes of this document are:
+
+- acts as a service discovery layer, allowing to get instances of a module from within another
+- provides access to messaging functionality (EventBus)
+
+### P2P - Persistence
+
+#### P2P AddressBook management / Network Churn
+
+Regardless of the underlying algorithm, `P2P` is a stateful business, there's state that has to be stored in memory (RAM) but there's also state that has to be persisted on the blockchain.
```
Comment (Olshansk) [inline integrations/README.md:61]: https://twitter.com/olshansky/status/1582506629354115072

File: integrations/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+In particular, when it comes `Network Churn` where `Actors` such as `Validators` are involved, joining and leaving the network, P2P has to integrate with Persistence in order to keep track of which actors were part of the network at that point in time (`Height` in Protocol terms).
+
+The running instance of the P2P module in a node, keeps in memory an "Address List" of the peers and that's being updated constantly.
```
Comment (Olshansk) [inline integrations/README.md:65]:
```suggestion
The running instance of the P2P module in a node, keeps an in memory "Address List" of the peers that's being constantly updated.
```

File: integrations/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+- acts as a service discovery layer, allowing to get instances of a module from within another
+- provides access to messaging functionality (EventBus)
+
+### P2P - Persistence
+
+#### P2P AddressBook management / Network Churn
+
+Regardless of the underlying algorithm, `P2P` is a stateful business, there's state that has to be stored in memory (RAM) but there's also state that has to be persisted on the blockchain.
+
+In particular, when it comes `Network Churn` where `Actors` such as `Validators` are involved, joining and leaving the network, P2P has to integrate with Persistence in order to keep track of which actors were part of the network at that point in time (`Height` in Protocol terms).
```
Comment (Olshansk) [inline integrations/README.md:63]: NIT: Should we backticket `P2P` and `Persistence`?

File: integrations/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+The running instance of the P2P module in a node, keeps in memory an "Address List" of the peers and that's being updated constantly.
+
+Whenever the `Consensus` determines a new `Height`, an event via the `Bus` notifies its listeners (`P2P` included) that the list of the current `Actors` that are part of the network has to be stored in `Persistence`.
```
Comment (Olshansk) [inline integrations/README.md:67]:
```suggestion
Whenever the `Consensus` module determines a new `Height`, an event via the `Bus` notifies its listeners (`P2P` included) that the list of the current `Actors` that are part of the network has to be updated at the `Persistence` layer.
```

**Review by Olshansk — CHANGES_REQUESTED** [review]
Few minor comments, but in general, I will contribute to this as well.

**Comment by Olshansk** [comment]
@deblasis Should we finish this off?

File: integrations/README.md (hunk 1 above)
Comment (deblasis) [reply integrations/README.md:61]: LOL

File: integrations/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
+# Pocket Network 1.0 Cross Module Integrations
+
+<table style="text-align: center;">
```
[the remark this answers is not in the document]
Comment (deblasis) [reply integrations/README.md:3]: 💯 
I think I copied it from somewhere else in the repo, it has been a while. I'll change it anyway.

Re: the link you pasted, I am not sure what you are pointing at, I see no tables, only a nice mermaid diagram. 

Regardless, I'll figure out :)

File: integrations/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+<tr style="text-align: center;">
+  <td style="border-collapse: collapse;">
+    <p align="center">
+        Alessandro De Blasis
```
[the remark this answers is not in the document]
Comment (deblasis) [reply integrations/README.md:7]: I assumed that was the logic. Should I change it?

[the remark this answers is not in the document]
Comment (deblasis) [reply integrations/README.md]: I changed it, for consistency I looked at other specs and I used the same format that puts the handle before the name

File: integrations/README.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+The purpose of this document is not to highlight the capabilities of each model but to focus on the point of contact/interactions with a brief explanation about the motivation behind the integration under discussion.
+
+## Integrations
+
+### Bus
+
+An application `Bus` serves as a cross-cutting integration medium.
+Without going too much into technical details, its main features for the purposes of this document are:
+
+- acts as a service discovery layer, allowing to get instances of a module from within another
```
[the remark this answers is not in the document]
Comment (deblasis) [reply integrations/README.md:27]: I see your points but perhaps this document should be less "loaded" with technicalities and try to explain things in simpler terms, good enough to have an idea of how it works at a high level

Regarding the diagram, if we copy-pasta it, we would have to keep track of it and keep them aligned.

I am sure that whoever reads this and is interested in how it works under the hood, they already pulled the main repo, probably as a first step actually.

Lemme know where your mind is at. Maybe we could ask Jess' point of view on this. I guess her input would be valuable. (not tagging her on purpose in the meantime).

My 2c. Let's keep it simple, actually, I was thinking that it might sound even too technical already perhaps 🤔

File: integrations/README.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+- provides access to messaging functionality (EventBus)
+
+### P2P - Persistence
+
+#### P2P AddressBook management / Network Churn
+
+Regardless of the underlying algorithm, `P2P` is a stateful business, there's state that has to be stored in memory (RAM) but there's also state that has to be persisted on the blockchain.
+
+In particular, when it comes `Network Churn` where `Actors` such as `Validators` are involved, joining and leaving the network, P2P has to integrate with Persistence in order to keep track of which actors were part of the network at that point in time (`Height` in Protocol terms).
```
Comment (deblasis) [inline integrations/README.md:36]:
```suggestion
In particular, when it comes `Network Churn` where `Actors` such as `Validators` are involved, joining and leaving the network, [P2P](https://github.com/pokt-network/pocket-network-protocol/blob/main/p2p/README.md) has to integrate with [Persistence](https://github.com/pokt-network/pocket-network-protocol/blob/main/persistence/README.md) in order to keep track of which actors were part of the network at that point in time (`Height` in Protocol terms).
```

**Comment by deblasis** [comment]
> @deblasis Should we finish this off?

@Olshansk PTAL

File: integrations/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply integrations/README.md:27]: I agree. This whole thing spec needs a revamp and I personally think https://github.com/pokt-network/pocket/issues/438 does a better job at introducing the P2P module. 

There will be more changes as we also have decided to adopt LibP2P.

- I've done an initial "revamp" of persistence (but will need another one in the future)
- Utility is in progress
- @bryanchriswhite should be able to take this one on over time
- Looking at @gokutheengineer for consensus

After we launch V1 (and have real data), we can write a single whitepaper capturing everything and publish on arxiv :)

**Review by Olshansk — APPROVED** [review]
